### PR TITLE
feat(core): dynamic runtime capability

### PR DESCRIPTION
**Pull request by lucasfernog**
Some capabilities must be dynamic (e.g. allow a dynamically started localhost server, runtime-defined remote URLs, fine-grained permissions that depend on user configuration like access token scopes..). This can be done already by creating a string at runtime and leaking it so it becomes a &'static str, but it's not ideal so we now support both runtime strings and added a new capability builder aswell so you can either `include_str!()` a static capability file or create/modify one at runtime.

### CODE REVIEW

File: core/tauri/src/ipc/authority.rs [hunk 1]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+      .get_or_insert_with(Default::default)
+      .urls
+      .push(url);
+    self
+  }
+
+  /// Do not apply this capability on local app URLs.
+  pub fn skip_local(mut self) -> Self {
+    self.0.local = false;
+    self
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs]: Probably better to just call it local and take a bool?
```suggestion
  pub fn local(mut self, local: bool) -> Self {
    self.0.local = local;
    self
```

File: core/tauri/src/ipc/authority.rs [hunk 2]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+  pub fn window(mut self, window: impl Into<String>) -> Self {
+    self.0.windows.push(window.into());
+    self
+  }
+
+  /// Link this capability to the given webview label.
+  pub fn webview(mut self, webview: impl Into<String>) -> Self {
+    self.0.webviews.push(webview.into());
+    self
+  }
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs]: would it be worth to also add iterator version of these two APIs as well?

File: core/tauri/src/ipc/authority.rs [hunk 3]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+    mut self,
+    permission: impl Into<String>,
+    allowed: Vec<T>,
+    denied: Vec<T>,
+  ) -> Self {
+    let permission = permission.into();
+    let identifier = permission
+      .clone()
+      .try_into()
+      .unwrap_or_else(|_| panic!("invalid permission identifier '{permission}'"));
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs:165]: I would try to avoid panicking and return a result instead

File: core/tauri/src/ipc/authority.rs [hunk 4]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+      .permissions
+      .push(PermissionEntry::ExtendedPermission {
+        identifier,
+        scope: Scopes {
+          allow: Some(
+            allowed
+              .into_iter()
+              .map(|a| {
+                serde_json::to_value(a)
+                  .expect("failed to serialize scope")
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs]: ditto

File: core/tauri/src/ipc/authority.rs [hunk 5]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+                  .into()
+              })
+              .collect(),
+          ),
+          deny: Some(
+            denied
+              .into_iter()
+              .map(|a| {
+                serde_json::to_value(a)
+                  .expect("failed to serialize scope")
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs]: ditto

File: core/tauri/src/ipc/authority.rs [hunk 6]
```diff
@@ -62,6 +65,128 @@ impl Origin {
...
+                  .into()
+              })
+              .collect(),
+          ),
+        },
+      });
+    self
```
Comment (amrbashir) [inline core/tauri/src/ipc/authority.rs]: let's spread these onto multiple lines to avoid crazy indentation

**Review by amrbashir — CHANGES_REQUESTED** [review]

**Comment by lucasfernog** [comment]
@amrbashir it's quite rare for those functions to panic (most identifiers are valid, and it's also almost impossible for serde_json::to_value to fail) and I don't really like the pattern of returning Result in a builder
plus we're also panicking on add_capability itself which I think is ok since if you mess up in the capability tou shouldn't ship your app at all

**Comment by amrbashir** [comment]
That's fair I guess

**Review by amrbashir — APPROVED** [review]

**Pull request merged**